packages/loader/container-loader/src/connectionManager.ts

```diff
@@ -891,20 +891,31 @@ export class ConnectionManager implements IConnectionManager {
 
         this._outbound.push(messages);
     }
-
+    private count: number = 0;
+    private readonly ops: any[] = [];
     public beforeProcessingIncomingOp(message: ISequencedDocumentMessage) {
         // if we have connection, and message is local, then we better treat is as local!
         assert(this.clientId !== message.clientId || this.lastSubmittedClientId === message.clientId,
             0x0ee /* "Not accounting local messages correctly" */,
         );
-
+            this.ops.push(message);
         if (this.lastSubmittedClientId !== undefined && this.lastSubmittedClientId === message.clientId) {
             const clientSequenceNumber = message.clientSequenceNumber;
 
             assert(this.clientSequenceNumberObserved < clientSequenceNumber, 0x0ef /* "client seq# not growing" */);
             assert(clientSequenceNumber <= this.clientSequenceNumber,
                 0x0f0 /* "Incoming local client seq# > generated by this client" */);
-
+                if (message.type !== "op") {
+                    this.count++;
+                }
+                const clientGap = clientSequenceNumber - this.clientSequenceNumberObserved - 1;
+                if (clientGap > 0) {
+                    if (this.count > 0) {
+                        this.count -= clientGap;
+                    } else {
+                        throw new Error("gap in client sequence #");
+                    }
```

Comment:
If I got this logic right, it finds a gap in clientSequence numbers. But then we substract size of that gap from number of all non-runtime ops. I'm not sure what this gives us (i.e. what `counter` is about)?

Reply:
i thought we are having a counter for noops like in this comment
"If this counter is above zero, and we have a gap, decrease counter by size of gap"
i'm doing a counter of not just noops but any non-runtime ops

Reply:
just counting the noops is not enough to account for the gap, the sharedTree fuzz tests were the reason i switched to tracking only runtime ops.